File: src/executors_py/executors_py/number_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.executors import SingleThreadedExecutor
from example_interfaces.msg import Int64

class NumberPublisherNode(Node):
    def __init__(self):
        super().__init__("number_publisher")
        self.number_ = 1
        self.publish_frequency_ = 1.0
        self.number_publisher_ = self.create_publisher(Int64, "number", 10)
        self.number_timer_ = self.create_timer(
            1.0 / self.publish_frequency_, self.publish_number)
        self.get_logger().info("Number publisher has been started.")

        # Main callbacks in ROS2:
        # - timers
        # - subscribers
        # - service servers
        # - action servers
        # - futures (in clients)


    def publish_number(self):
        msg = Int64()
        msg.data = self.number_
        self.number_publisher_.publish(msg)
        self.number_ += 1

def main(args=None):
    rclpy.init(args=args)
    node = NumberPublisherNode()
    
    executor = SingleThreadedExecutor()
    executor.add_node(node) # Add the node to the executor
    executor.spin() # Keep spinning the executor

    rclpy.shutdown()

File: src/components_cpp/src/number_publisher.cpp
#include "components_cpp/number_publisher.hpp"

using namespace std::chrono_literals;

namespace my_namespace
{
    NumberPublisher::NumberPublisher(const rclcpp::NodeOptions& options) : Node("number_publisher", options)
    {
        number_ = 2;

        number_publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number", 10);
        number_timer_ = this->create_wall_timer(1000ms,
                                                std::bind(&NumberPublisher::publishNumber, this));
        RCLCPP_INFO(this->get_logger(), "Number publisher has been started.");
    }
    void NumberPublisher::publishNumber()
    {
        auto msg = example_interfaces::msg::Int64();
        msg.data = number_;
        number_publisher_->publish(msg);
    }
} // namespace my_namespace

#include "rclcpp_components/register_node_macro.hpp"
RCLCPP_COMPONENTS_REGISTER_NODE(my_namespace::NumberPublisher)
